### PR TITLE
Reconcile testing docs, fixes #18606

### LINKED ISSUE #18606: Reconcile testing docs
https://github.com/kubernetes/kubernetes/pull/14112 introduced [docs/devel/e2e-tests.md](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/e2e-tests.md), but that document isn't correct, and contradicts the information in [docs/devel/development.md](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md).  These two docs should be reconciled.

### PR DESCRIPTION
**Pull request by ikehz**
cc @kubernetes/goog-testing @kubernetes/sig-testing

### CODE REVIEW

**Comment by ihmccreery** [comment]
Also cc @quinton-hoole @maisem

File: docs/devel/e2e.md [hunk 1]
```diff
@@ -84,23 +143,40 @@ Prior to running the tests, it is recommended that you first create a simple aut
...
 
-As mentioned earlier there are a host of other options that are available, but are left to the developer
+This will start a single-node Kubernetes cluster than runs pods using the local docker daemon.  Press Control-C to stop the cluster.
 
-**NOTE:** If you are running tests on a local cluster repeatedly, you may need to periodically perform some manual cleanup.
-- `rm -rf /var/run/kubernetes`, clear kube generated credentials, sometimes stale permissions can cause problems.
-- `sudo iptables -F`, clear ip tables rules left by the kube-proxy.
+#### Testing against local clusters
+
+In order to run an E2E test against a locally running cluster, use the `e2e.test` binary built by `hack/build-go.sh`
```
Comment (ixdy) [inline docs/devel/e2e.md]: this contradicts the example given

File: docs/devel/e2e.md [hunk 2]
```diff
@@ -32,33 +32,92 @@ Documentation for other releases can be found at
...
-`$ make all`
+```sh
+go run hack/e2e.go -v --build --up --test --down
+```
+
+If you'd like to just perform one of these steps, here are some examples:
+
+```sh
+# Build binaries for testing
+go run hack/e2e.go --build
```
Comment (spxtr) [inline docs/devel/e2e.md]: `-v` for these, or else you don't get any helpful output when they fail.

File: docs/devel/e2e.md [hunk 3]
```diff
@@ -32,33 +32,92 @@ Documentation for other releases can be found at
...
+
+# Create a fresh cluster.  Deletes a cluster first, if it exists
+go run hack/e2e.go --up
```
Comment (spiffxp) [inline docs/devel/e2e.md]: Modify one of these examples to demonstrate a non-default provider?  eg:
```
# Crate a fresh cluster using the 'aws' provider.  Deletes a cluster first, if it exists
KUBERNETES_PROVIDER=aws go run hack/e2e.go --up`

File: docs/devel/e2e.md [hunk 4]
```diff
@@ -116,7 +192,13 @@ We are working on implementing clearer partitioning of our e2e tests to make run
...
   - If a feature is in beta or GA, it *should* block the merge-queue.  In moving from experimental to beta or GA, tests that are expected to pass by default should simply remove the `[Feature:...]` label, and will be incorporated into our core suites.  If tests are not expected to pass by default, (e.g. they require a special environment such as added quota,) they should remain with the `[Feature:...]` label, and the suites that run them should be incorporated into our merge-queue, owned by the Build Cop.
 
-Finally, `[Conformance]` tests are tests we expect to pass on **any** Kubernetes cluster.  The `[Conformance]` label does not supersede any other labels.  `[Conformance]` test policies are a work-in-progress; see #18162.
+### Conformance tests
+
+Finally, `[Conformance]` tests are tests we expect to pass on **any** Kubernetes cluster.  The `[Conformance]` label does not supersede any other labels.  `[Conformance]` test policies are a work-in-progress (see #18162).
+
+End-to-end testing, as described above, is for [development distributions](writing-a-getting-started-guide.md).  A conformance test is used on a [versioned distro](writing-a-getting-started-guide.md).
```
Comment (spiffxp) [inline docs/devel/e2e.md]: links TBD

File: docs/devel/e2e.md [hunk 5]
```diff
@@ -32,33 +32,92 @@ Documentation for other releases can be found at
...
+End-to-end (e2e) tests for Kubernetes provide a mechanism to test end-to-end behavior of the system, and is the last signal to ensure end user operations match developer specifications.  Although unit and integration tests should ideally provide a good signal, the reality is in a distributed system like Kubernetes it is not uncommon that a minor change may pass all unit and integration tests, but cause unforseen changes at the system level.  e2e testing is very costly, though: it takes a long time to build, deploy, and exercise a cluster.  Thus, the primary objectives of the e2e tests are to ensure a consistent and reliable behavior of the kubernetes code base, and to catch hard-to-test bugs before users do, when unit and integration tests are insufficient.
```
Comment (timothysc) [inline docs/devel/e2e.md]: s/e2e testing is very costly/ something quantifiable.. 

costly in what?  time?

File: docs/devel/e2e.md [hunk 6]
```diff
@@ -32,33 +32,92 @@ Documentation for other releases can be found at
...
+The e2e tests in kubernetes are built atop of [Ginkgo](http://onsi.github.io/ginkgo/) and [Gomega](http://onsi.github.io/gomega/).  There are a host of features that this BDD testing framework provides, and it is recommended that the developer read the documentation prior to diving into the tests.
 
-The purpose of *this* document is to serve as a primer for developers who are looking to execute, or add tests, using a local development environment.
+The purpose of *this* document is to serve as a primer for developers who are looking to execute or add tests using a local development environment.
 
 ## Building and Running the Tests
 
-**NOTE:** The tests have an array of options.  For simplicity, the examples will focus on leveraging the tests on a local cluster using `sudo ./hack/local-up-cluster.sh`
+There are a variety of ways to run e2e tests, but we aim to decrease the number of ways to run e2e tests to a canonical way: `hack/e2e.go`.
```
Comment (timothysc) [inline docs/devel/e2e.md]: Have you verified that it all works in a non-GCE env.  Many moons ago it had not.

File: docs/devel/e2e.md [hunk 7]
```diff
@@ -32,33 +32,92 @@ Documentation for other releases can be found at
...
 
-### Building the Tests
+You can run an end-to-end test which will bring up a master and nodes, perform some tests, and then tear everything down. Make sure you have followed the getting started steps for your chosen cloud platform (which might involve changing the `KUBERNETES_PROVIDER` environment variable to something other than "gce").
```
Comment (timothysc) [inline docs/devel/e2e.md]: This is a little too non-descriptive.  Isn't ./hack always 'local'?

File: docs/devel/e2e.md [hunk 8]
```diff
@@ -32,33 +32,92 @@ Documentation for other releases can be found at
...
 
-The tests are built into a single binary which can be run against any deployed kubernetes system.  To build the tests, navigate to your source directory and execute:
+To build Kubernetes, up a cluster, run tests, and tear everything down, use:
 
-`$ make all`
+```sh
+go run hack/e2e.go -v --build --up --test --down
```
Comment (timothysc) [inline docs/devel/e2e.md]: If we're going to explain this we need to outline the flags.

File: docs/devel/e2e.md [hunk 9]
```diff
@@ -32,33 +32,92 @@ Documentation for other releases can be found at
...
+
+# Push to an existing cluster, or bring up a cluster if it's down.
+go run hack/e2e.go --pushup
+
+# Run all tests
+go run hack/e2e.go --test
 
-The output for the end-2-end tests will be a single binary called `e2e.test` under the default output directory, which is typically `_output/local/bin/linux/amd64/`.  Within the repository there are scripts that are provided under the `./hack` directory that are helpful for automation, but may not apply for a local development purposes.  Instead, we recommend familiarizing yourself with the executable options.  To obtain the full list of options, run the following:
+# Run tests matching the regex "\[Conformance\]" (the conformance tests)
+go run hack/e2e.go -v -test --test_args="--ginkgo.focus=\[Conformance\]"
```
Comment (timothysc) [inline docs/devel/e2e.md]: ug.. we need to manually escape vs. going to the tests themselves?  

Why would we be pointing them to hack/e2e.go vs. just pointing them to ginkgo/e2e.test directly?  I really don't see why this is easier, or better...?

File: docs/devel/e2e.md [hunk 10]
```diff
@@ -32,33 +32,92 @@ Documentation for other releases can be found at
...
+#
+# e.g.:
+go run hack/e2e.go -v --build --pushup --test --down
 
-For the purposes of brevity, we will look at a subset of the options, which are listed below:
+# -ctl can be used to quickly call kubectl against your e2e cluster. Useful for
+# cleaning up after a failed test or viewing logs. Use -v to avoid suppressing
+# kubectl output.
+go run hack/e2e.go -v -ctl='get events'
+go run hack/e2e.go -v -ctl='delete pod foobar'
```
Comment (timothysc) [inline docs/devel/e2e.md]: why would I ever do this vs. using the explicit tooling?

**Comment by timothysc** [comment]
So I have to admit, I'm confused by which tool we should be recommending for what operations?  It seems like the e2e.go tries to be all things, but why would I ever use it in that way vs. using the underlying tools that provide the actual behavior?  

/cc @spiffxp

File: docs/devel/e2e.md (hunk 7 above)
Comment (ihmccreery) [inline docs/devel/e2e.md]: Not sure what you mean?  `./hack/e2e.go` is by default run on GCE.

File: docs/devel/e2e.md (hunk 8 above)
Comment (ihmccreery) [inline docs/devel/e2e.md]: There are explanations later on; is that okay?

File: docs/devel/e2e.md (hunk 6 above)
Comment (ihmccreery) [inline docs/devel/e2e.md]: Haven't verified; opened #20861.

File: docs/devel/e2e.md (hunk 9 above)
Comment (ihmccreery) [inline docs/devel/e2e.md]: See below.

**Comment by ihmccreery** [comment]
> Why would we be pointing them to hack/e2e.go vs. just pointing them to ginkgo/e2e.test directly? I really don't see why this is easier, or better...?

@timothysc I totally agree with you, but right now our CI infrastructure uses `e2e.go`, and I want to make sure that devs know how to replicate results locally that we're getting in CI.  We've gotten feedback from @kubernetes/sig-testing that there should be a canonical way to run tests, and in this document I'm trying to reflect reality.

File: docs/devel/e2e.md (hunk 10 above)
Comment (ihmccreery) [inline docs/devel/e2e.md]: Not sure what you mean by 'explicit tooling'?  (Perhaps see comment below?)

**Comment by ihmccreery** [comment]
@timothysc PTAL.

File: docs/devel/e2e.md (hunk 4 above)
Comment (ihmccreery) [inline docs/devel/e2e.md]: Done.

File: docs/devel/e2e.md (hunk 10 above)
Comment (timothysc) [inline docs/devel/e2e.md]: > Not sure what you mean by 'explicit tooling'?

why wouldn't I just use ```kubectl get events``` etc.

Comment (ihmccreery) [inline docs/devel/e2e.md]: Oh; if you have a system installation, `kubectl` will route to that, not the locally-built one.  This is a shortcut to `./cluster/kubectl.sh` IIUC.  Why, though, I couldn't tell you.  I'm just munging docs.  :stuck_out_tongue:

**Comment by spiffxp** [comment]
Overall LGTM as a reflection of what is done today; I agree with @timothysc that the `ctl` command doesn't make much sense to me.  

Path forward is to probably tear apart the `isup/up/down/push` functionality from `test` functionality, to start exposing all the magic filenames, keynames, environment variables, etc. that are transparently getting put into place today.  Great candidate for v1.3.

**Comment by ihmccreery** [comment]
@timothysc Ping.

**Comment by timothysc** [comment]
LGTM.

**Pull request merged**